src/Bicep.Core/Intermediate/ExpressionBuilder.cs

```diff
@@ -271,6 +271,11 @@ private TypeExpression ConvertTypeWithoutLowering(SyntaxBase syntax)
                 GetTypeInfo<ArrayType>(syntax),
                 ConvertTypeWithoutLowering(arrayTypeSyntax.Item.Value)),
             NullableTypeSyntax nullableTypeSyntax => new NullableTypeExpression(syntax, ConvertTypeWithoutLowering(nullableTypeSyntax.Base)),
+            UnionTypeSyntax unionTypeSyntax when GetDiscriminatedObjectUnionFromTypeInfo(Context.SemanticModel.GetTypeInfo(unionTypeSyntax)) is { } discriminatedObjectType =>
```

Comment:
I think you can simplify to:

```suggestion
            UnionTypeSyntax unionTypeSyntax when Context.SemanticModel.GetTypeInfo(unionTypeSyntax) is DiscriminatedObjectType discriminatedObjectType =>
```